### PR TITLE
Custom headers by service labels for docker backends

### PR DESCRIPTION
**Pull request by Tiscs**
### What does this PR do?

Supports to configured custom request/response headers by service labels for docker backends.

### Motivation

Allowed to set custom headers for special endpoint for single docker backend.

### More

- [x] Added/updated tests
- [x] Added/updated documentation

### CODE REVIEW

**Comment by dtomcej** [comment]
This will need a rebase, since the 1.5 merge to master!

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
 /dist
 /autogen/genstatic/gen.go
+/autogen/gentemplates/gen.go
```
Comment (dtomcej) [inline .gitignore]: Not sure this is right.

File: provider/docker/labels.go [hunk 2]
```diff
@@ -22,6 +22,23 @@ const (
 // we can get it with label[serviceName][propertyName] and we got the propertyValue
 type labelServiceProperties map[string]map[string]string
 
+// Common functions
+
+func parseMapString(value string) map[string]string {
```
Comment (dtomcej) [inline provider/docker/labels.go]: This function is being changed. Will need to update this one.

File: .gitignore (hunk 1 above)
Comment (Tiscs) [reply .gitignore]: I think the file `/autogen/gentemplates/gen.go` should be generate when build, likes `/autogen/gen.go` in **v1.4**.

**Comment by Tiscs** [comment]
@ldez I merged branch `origin/master` yet.
May be I can close this PR and create a new PR with an other branch.

File: provider/docker/config_service_test.go [hunk 3]
```diff
@@ -9,6 +9,63 @@ import (
 	docker "github.com/docker/docker/api/types"
 )
 
+func TestDockerGetFuncMapLabel(t *testing.T) {
```
Comment (mmatur) [inline provider/docker/config_service_test.go]: Could you please improve your test to add cases with mutliple custom request/response headers ?

File: provider/docker/config_service.go [hunk 4]
```diff
@@ -114,6 +120,13 @@ func hasServiceLabel(container dockerData, serviceName string, labelSuffix strin
 	return label.Has(container.Labels, label.Prefix+labelSuffix)
 }
 
+func getServiceMapLabel(container dockerData, serviceName string, labelSuffix string) map[string]string {
+	if value, ok := getServiceLabels(container, serviceName)[labelSuffix]; ok {
+		return label.ParseMapValue(label.Prefix+serviceName+"."+labelSuffix, value)
```
Comment (ldez) [inline provider/docker/config_service.go]: Could can rewrite like that:
```go
lblName := label.GetServiceLabel(labelSuffix, serviceName)
return label.ParseMapValue(lblName, value)
```

File: provider/docker/config_service_test.go [hunk 5]
```diff
@@ -9,6 +9,63 @@ import (
...
+			container: containerJSON(labels(map[string]string{
+				label.TraefikFrontendRequestHeaders: "X-Custom-Header: TEST",
+			})),
+			suffixLabel:   label.SuffixFrontendRequestHeaders,
+			expectedKey:   "X-Custom-Header",
+			expectedValue: "TEST",
+		},
+		{
+			container: containerJSON(labels(map[string]string{
+				label.Prefix + serviceName + "." + label.SuffixFrontendRequestHeaders: "X-Custom-Header: TEST",
```
Comment (ldez) [inline provider/docker/config_service_test.go]:
```go
label.GetServiceLabel(labelSuffix, label.SuffixFrontendRequestHeaders)
```

File: provider/docker/config_service_test.go [hunk 6]
```diff
@@ -9,6 +9,63 @@ import (
...
+			container: containerJSON(labels(map[string]string{
+				label.TraefikFrontendResponseHeaders: "X-Custom-Header: TEST",
+			})),
+			suffixLabel:   label.SuffixFrontendResponseHeaders,
+			expectedKey:   "X-Custom-Header",
+			expectedValue: "TEST",
+		},
+		{
+			container: containerJSON(labels(map[string]string{
+				label.Prefix + serviceName + "." + label.SuffixFrontendResponseHeaders: "X-Custom-Header: TEST",
```
Comment (ldez) [inline provider/docker/config_service_test.go]:
```go
label.GetServiceLabel(labelSuffix, label.SuffixFrontendResponseHeaders)
```

File: provider/docker/config_service_test.go (hunk 3 above)
Comment (Tiscs) [reply provider/docker/config_service_test.go]: @mmatur Done

**Pull request merged**